### PR TITLE
Adding `jill` from PyPI

### PR DESCRIPTION
**Pull request by sugatoray**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by sugatoray** [comment]
## Build Failure Reason

`jill 0.10.1 requires wget, which is not installed.` [(error)](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=446070&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823&l=1261)

```sh
import: 'jill'
+ pip check
jill 0.10.1 requires wget, which is not installed.
Tests failed for jill-0.10.1-pyhd8ed1ab_0.tar.bz2 - moving package to /home/conda/staged-recipes/build_artifacts/broken
WARNING:conda_build.build:Tests failed for jill-0.10.1-pyhd8ed1ab_0.tar.bz2 - moving package to /home/conda/staged-recipes/build_artifacts/broken
WARNING conda_build.build:tests_failed(2970): Tests failed for jill-0.10.1-pyhd8ed1ab_0.tar.bz2 - moving package to /home/conda/staged-recipes/build_artifacts/broken
TESTS FAILED: jill-0.10.1-pyhd8ed1ab_0.tar.bz2
```

**Comment by sugatoray** [comment]
PR is ready. Request for review/merging.

cc: @conda-forge/help-python @conda-forge/staged-recipes

File: recipes/jill/meta.yaml
Comment (BastianZim) [inline recipes/jill/meta.yaml:25]: Why is all of this needed?

File: recipes/jill/meta.yaml
Comment (BastianZim) [inline recipes/jill/meta.yaml:11]: Why is this needed? The file should be included in the package. If not, the source should probably be changed to GitHub and upstream notified.

File: recipes/jill/meta.yaml
Comment (BastianZim) [inline recipes/jill/meta.yaml:32]: Is this needed in host? If pip check fails it's better to either patch it so `wget` is not checked or remove pip check.

File: recipes/jill/meta.yaml
Comment (BastianZim) [inline recipes/jill/meta.yaml:52]: Are you sure? Anything under run is available during testing.

File: recipes/jill/yum_requirements.txt
Comment (BastianZim) [inline recipes/jill/yum_requirements.txt:1]: Is this really required, or did you only include it for pip check?

**Review by BastianZim — COMMENTED** [review]
Thanks for adding – Have added a couple of comments.

Please only ping the relevant review team in the future, not staged-recipes and gitter as well. If you don't get a review, feel free to ask there but only after some time.
Otherwise, a lot of noise is introduced.

**Comment by sugatoray** [comment]
Thank you @BastianZim. Recently I saw a few other PRs doing that and wondered perhaps it's okay to do that. Apologies for any unintended noise created for this. 

A question: when is it okay to ping the staging team?

And when I leave a note in the gitter chat, am I reaching out to the core team (I see some of the core team members there often: `@beckermr`, `@jaimergp`, `@dopplershit`)?

File: recipes/jill/meta.yaml
Comment (sugatoray) [reply recipes/jill/meta.yaml:52]: Had tried with it uncommented. But it was failing. It did not matter where I added wget. It would still fail in the pip-check step. So, got it commented to pass it. 

But did not remove any of the comments, only to see if the review provides some better alternatives. 

Just to be clear, in the worse scenario, I will remove all inclusions of wget from the recipe and remove pip-check as well.

File: recipes/jill/meta.yaml
Comment (sugatoray) [reply recipes/jill/meta.yaml:32]: Have not had created a patch yet. Have seen some documentation on patching though. But I don't have clarity on what goes on technically when a patch is used. Could you please elaborate on that? (perhaps with an example will be very helpful).

Thank you.

File: recipes/jill/meta.yaml
Comment (sugatoray) [reply recipes/jill/meta.yaml:11]: Yes, that will be the best solution: fixing it at the source. But before I do that, I wanted to see if that's the only thing missing. Left this commented as the source is being pulled from PyPI, pulling this file from the upstream-repo could make it pull a different version of the file. But, leaving it as a comment there helps as a reminder to raise an issue with upstream.

File: recipes/jill/yum_requirements.txt
Comment (sugatoray) [reply recipes/jill/yum_requirements.txt:1]: Only for pip-check, since including it in the meta.yaml file made no difference. But I don't think that this works either.

File: recipes/jill/meta.yaml
Comment (sugatoray) [reply recipes/jill/meta.yaml:25]: It is a temporary measure to make the recipe pass the build. I will open an issue at upstream to ask them to include the file in their next release. But, if anything else is missing, would be helpful to know that before I open the issue. Thus can include all of that in a single issue.

Comment (BastianZim) [reply recipes/jill/meta.yaml:25]: Ok in that case it's better to switch to GitHub releases (If available) as these scripts don't always work and introduce unnecessary complexity. Conda-forge currently has enough CI time so no problem if the package is slightly bigger.

File: recipes/jill/meta.yaml
Comment (BastianZim) [reply recipes/jill/meta.yaml:11]: Ahh ok. Then I would just raise the issue now and remove this line as it's cleaner that way. Conda-forge recipes should be as clean as possible as it sometimes can confuse the bots or introduce other overlooked points of failure – just as a best practice.

File: recipes/jill/meta.yaml
Comment (BastianZim) [reply recipes/jill/meta.yaml:32]: Host are packages available during build and is primarily important when building against something or when cross-compiling so not this situation.

Pip check just checks if all packages can be discovered by pip. This requires, IIRC, that the package was built with pip in the first place which is not always the case. So pip check can fail even if the package is installed correctly. This also explains why you can import it but the check still fails.

A patch is just a normal git patch that is applied to your source: https://git-scm.com/docs/git-format-patch
But the easiest approach is probably to just remove pip check here. it's not mandatory to have it here.

File: recipes/jill/meta.yaml
Comment (BastianZim) [reply recipes/jill/meta.yaml:52]: 😄 

Ahh ok. I would suggest adding all of the necessary requirements specified by upstream (If that includes wget, it should be added as well) and then removing pip check.

File: recipes/jill/yum_requirements.txt
Comment (BastianZim) [reply recipes/jill/yum_requirements.txt:1]: No this is just pip check not recognising it. Just remove it from here and list it under `run`– that should be enough.

**Comment by BastianZim** [comment]
No worries. Yes, others use it but it's not encouraged there either.

Generally, the procedure should be:

1. contact relevant review team or staged recipes if no review team exists for your language
2. Wait a couple of days and if no reply is received ping the staged recipes team
3. If still no reply, ask in Gitter.

Most don't really adhere to this and if your package is urgently needed you can also post in Gitter directly but that's best practise.

**Comment by sugatoray** [comment]
Thank you, for explaining. That helps.

**Comment by sugatoray** [comment]
Since the azure CI's are failing at the moment, I am not making any changes to the recipe. I will make the suggested changes, once the CI's are back online.

**Comment by BastianZim** [comment]
CIs are back up :tada:

**Comment by sugatoray** [comment]
@BastianZim I opened the following issue to include `requirements.txt` in upstream source.

- https://github.com/johnnychen94/jill.py/issues/110

Also, I have removed pip-check and othet lines that are not necessary.

See raw: [`requirements.txt`](https://raw.githubusercontent.com/johnnychen94/jill.py/master/requirements.txt)

**Comment by johnnychen94** [comment]
@sugatoray Thanks for the efforts! jill 0.11.0 is out: https://pypi.org/project/jill/

**Comment by johnnychen94** [comment]
Will it be easier to maintain the conda-forge build recipe if I switch from the plain `requirements.txt` to the fancy poetry `pyproject.toml` and replace the `wget` dependency with `requests`?

**Comment by BastianZim** [comment]
@johnnychen94 Thanks for the offer but no that's fine. The package just needs to be built from source. As long as that's working we're good.

**Comment by BastianZim** [comment]
@sugatoray Ready for review/merge or is there anything else you would like to update?

**Comment by sugatoray** [comment]
@BastianZim I think we are good here. It is ready for review. Thank you, for the detailed review so far.

File: recipes/jill/meta.yaml
Comment (BastianZim) [inline recipes/jill/meta.yaml:23]:
```suggestion
```

File: recipes/jill/meta.yaml
Comment (BastianZim) [inline recipes/jill/meta.yaml:36]:
```suggestion
test:
  imports:
    - jill
  commands:
    - jill --help  
```
You can still test the command line.

**Review by BastianZim — CHANGES_REQUESTED** [review]
Just two small things and this should be good to go!

File: recipes/jill/meta.yaml
Comment (sugatoray) [reply recipes/jill/meta.yaml]: Ah. Forgot about that one. Thanks.

**Comment by sugatoray** [comment]
Done  ✅

**Comment by sugatoray** [comment]
PR is ready. Request for review/merging.

cc: @conda-forge/help-python

**Comment by sugatoray** [comment]
@BastianZim The line `jill --help` now throws an error.

```sh
ModuleNotFoundError: No module named 'wget'
```

### Details

[(error)](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=464220&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823&l=1345)

```sh
import: 'jill'
+ jill --help
Traceback (most recent call last):
  File "/home/conda/staged-recipes/build_artifacts/jill_1645647151740/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/bin/jill", line 6, in <module>
    from jill.__main__ import main
  File "/home/conda/staged-recipes/build_artifacts/jill_1645647151740/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/lib/python3.10/site-packages/jill/__main__.py", line 1, in <module>
    from .download import download_package
  File "/home/conda/staged-recipes/build_artifacts/jill_1645647151740/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/lib/python3.10/site-packages/jill/download.py", line 11, in <module>
    import wget
ModuleNotFoundError: No module named 'wget'

```

**Comment by BastianZim** [comment]
Yeah that is indeed strange. I'll have to have a more in-depth look at that.

Removing the tests is unfortunately not possible as the package has to work before I can merge it. :)

File: recipes/jill/meta.yaml
Comment (dopplershift) [inline recipes/jill/meta.yaml:31]: This is depending on the GNU `wget` command line utility, not the [PyPI](https://pypi.org/project/wget/) package, which has not yet been packaged for conda-forge.

Comment (sugatoray) [reply recipes/jill/meta.yaml:31]: @johnnychen94 Can you instead use the PyPI `wget` or `requests` as you were suggesting? Perhaps that will help circumvent this hurdle. 

@BastianZim @dopplershift Thank you, for looking into it.

Comment (dopplershift) [reply recipes/jill/meta.yaml:31]: @sugatoray To clarify: `jill` already depends on the PyPI `wget` package. The problem is that as written above, the recipe is requesting conda-forge's `wget` package, which downloads the commandline `wget` package, which is why `import wget` fails. I thought the PyPI version wasn't available, but I was wrong. This line just needs to change to:
```suggestion
    - python-wget
```

Comment (BastianZim) [reply recipes/jill/meta.yaml:31]: @sugatoray Just as a general FYI, Grayskull maintains a mapping of PyPI to CF names here: https://github.com/conda-incubator/grayskull/blob/main/grayskull/strategy/config.yaml

If you find these kinds of exceptions please feel free to add them there.

Comment (sugatoray) [reply recipes/jill/meta.yaml]: @BastianZim I got this pypi-cf name-mapping from `@beckermr` (as far as I recall).

- https://github.com/regro/cf-graph-countyfair/blob/master/mappings/pypi/grayskull_pypi_mapping.yaml

File: recipes/jill/meta.yaml
Comment (sugatoray) [reply recipes/jill/meta.yaml]: Since, pip check failed, I dropped it from the recipe.

**Comment by sugatoray** [comment]
> Yeah that is indeed strange. I'll have to have a more in-depth look at that.
> 
> Removing the tests is unfortunately not possible as the package has to work before I can merge it. :)

So, the pip check fails with `wget` or `python-wget`. And so does running the command `jill --help`. Any suggestions?

cc: @BastianZim @dopplershift

**Comment by sugatoray** [comment]
Keep it open

**Pull request closed without merging**
